### DISCUSSION TITLE
Waiting for load-balancer to get an IP...

### DISCUSSION THREAD

**Discussion by grig0701** [kube-hetzner/terraform-hcloud-kube-hetzner#922] [Q&A] [answered]
### Description

Hello. We planned to use this module for production and development environments, but during the dev phase, there was a problem with module.kube-hetzner.null_resource.kustomization. Everything worked fine until today. Please help us fix this. Can we do this without rebuilding the cluster?

I also have a question. Can we add other servers to the network that creates a module for itself, for example, a separate instance for Jenkins outside of the cluster? Will this not break the module?

```
Connecting to remote host via SSH...
    Host: [Our IP master]
    User: root
    Password: false
    Private key: true
    Certificate: false
    SSH Agent: false
    Checking Host Key: false
    Target Platform: unix
  Connected!
  + sed -i 's/^- |[0-9]\+$/- |/g' /var/post_install/kustomization.yaml
  + timeout 360 bash
  + kubectl apply -k /var/post_install
  namespace/cattle-system unchanged
  namespace/cert-manager unchanged
  namespace/nginx unchanged
  namespace/system-upgrade unchanged
  storageclass.storage.k8s.io/hcloud-volumes unchanged
  serviceaccount/hcloud-cloud-controller-manager unchanged
  serviceaccount/hcloud-csi-controller unchanged
  serviceaccount/kured unchanged
  serviceaccount/system-upgrade unchanged
  role.rbac.authorization.k8s.io/kured unchanged
  clusterrole.rbac.authorization.k8s.io/hcloud-csi-controller unchanged
  clusterrole.rbac.authorization.k8s.io/kured unchanged
  rolebinding.rbac.authorization.k8s.io/kured unchanged
  clusterrolebinding.rbac.authorization.k8s.io/hcloud-csi-controller unchanged
  clusterrolebinding.rbac.authorization.k8s.io/kured unchanged
  clusterrolebinding.rbac.authorization.k8s.io/system-upgrade unchanged
  clusterrolebinding.rbac.authorization.k8s.io/system:hcloud-cloud-controller-manager unchanged
  configmap/default-controller-env unchanged
  service/hcloud-csi-controller-metrics unchanged
  service/hcloud-csi-node-metrics unchanged
  deployment.apps/hcloud-cloud-controller-manager unchanged
  deployment.apps/hcloud-csi-controller unchanged
  deployment.apps/system-upgrade-controller unchanged
  daemonset.apps/hcloud-csi-node unchanged
  daemonset.apps/kured unchanged
  helmchart.helm.cattle.io/cert-manager unchanged
  helmchart.helm.cattle.io/nginx unchanged
  helmchart.helm.cattle.io/rancher unchanged
  csidriver.storage.k8s.io/csi.hetzner.cloud unchanged
  + echo 'Waiting for the system-upgrade-controller deployment to become available...'
  Waiting for the system-upgrade-controller deployment to become available...
  + kubectl -n system-upgrade wait --for=condition=available --timeout=360s deployment/system-upgrade-controller
  deployment.apps/system-upgrade-controller condition met
  + sleep 7
  + kubectl -n system-upgrade apply -f /var/post_install/plans.yaml
  plan.upgrade.cattle.io/k3s-agent unchanged
  plan.upgrade.cattle.io/k3s-server unchanged
  + timeout 360 bash
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
  Waiting for load-balancer to get an IP...
```

### Kube.tf file

```terraform
module "kube-hetzner" {

  providers = {
    hcloud = hcloud
  }

  hcloud_token         = var.hcloud_token
  source               = "kube-hetzner/kube-hetzner/hcloud"
  ssh_public_key       = var.pub
  ssh_private_key      = var.priv
  network_region       = "eu-central"
  create_kubeconfig    = false
  create_kustomization = false

  use_control_plane_lb  = true
  lb_hostname           = "hostname"
  control_plane_lb_type = "lb11"
  enable_rancher        = true
  initial_k3s_channel   = "v1.25"
  rancher_hostname      = "hostname"

  cluster_name = "k3s"

  automatically_upgrade_k3s = false
  automatically_upgrade_os  = false

  ingress_controller = "nginx"

  control_plane_nodepools = [
    {
      name        = "control-plane",
      server_type = "cx31",
      location    = "fsn1",
      labels      = ["role=master"],
      taints      = [],
      count       = 3

      # Enable automatic backups via Hetzner (default: false)
      # backups = true
    },

  ]

  agent_nodepools = [
    {
      name        = "agent-development",
      server_type = "cpx41",
      location    = "fsn1",
      labels      = ["environment=development", "role=worker"],
      taints      = [],
      count       = 2

      # Enable automatic backups via Hetzner (default: false)
      # backups = true
    },
    {
      name        = "agent-jenkins-main",
      server_type = "cx41",
      location    = "fsn1",
      labels      = ["environment=jenkins"],
      taints      = [],
      count       = 0

      # Enable automatic backups via Hetzner (default: false)
      # backups = true
    },
    {
      name        = "agent-redis-dev",
      server_type = "cx21",
      location    = "fsn1",
      labels      = ["environment=redis"],
      taints      = [],
      count       = 1

      # Enable automatic backups via Hetzner (default: false)
      # backups = true
    },
  ]

  load_balancer_type     = "lb11"
  load_balancer_location = "fsn1"

}
```

### Platform

Linux

**Comment by Keyruu** [comment]
Same issue for me

**Comment by mysticaltech** [comment]
@grig0701 @Keyruu  Sometimes the Hetzner systems are slow, you retry. And if it does not work, upgrade to the latest version, see readme. I just tested it works now.

@grig0701 Regarding your second question, of you could can deploy other instances, terraform makes this easy, as you can include our module in a greater terraform project.

**Reply by grig0701** [reply]
Hello. Thanks for the answer. I just updated the module to 2.5.0, but the problem is still not solved. module.kube-hetzner.null_resource.kustomization waits for the load balancer and then crashes with an error

**Comment by MoergJ** [comment]
I just had a quick look. When using nginx as the Ingress Controller, it states to use a Domain name for the main Load-Balancer, which results the service not to get assigned an IP address but the selected domain name. 
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/2cfcfc74cb30afe7e401668521876f98322953ac/init.tf#L296C217-L296C255
Here it checks for the IP explicitly, which is empty when using a domain, it shoud rather check the
`--output=jsonpath='{.status.loadBalancer.ingress[0].hostname}'`

**Reply by MoergJ** [reply]
A little further specification: In the readme, it says when using the Cert-Manager together wirh nginx Ingress-Controller and HTTP challenge, the variable `lb_hostname = "cluster.example.org"` should be used. This results in the domain name being assigned to the service rather than it's IP.

**Reply by M4t7e** [reply]
Related to https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/908

@mysticaltech It appears the revert was not fully done.
See: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/908#issuecomment-1657035380 and the corresponding commit  to revert https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/commit/4047ca19af19f4953087699ca953f48d5f908f6d

**Reply by mysticaltech** [reply]
@MoergJ  PR welcome for the fix, it would help a lot. But please test it before submitting. It issue was introduced by a supposed fix we trusted recently that turned out not great.

@M4t7e Yep, I had reversed half of it and assumed the other half was good! 🤦‍♂️

**Reply by M4t7e** [reply]
@mysticaltech Ah okay, I see. I could also reproduce the issue now and it really seems to block setting a `lb_hostname`.

I added #924 to revert the last piece.

**Comment by mysticaltech** [comment]
@MoergJ  @M4t7e Currently reverting #908 completely. And we will improve the logic from there in the future.

**Reply by M4t7e** [reply]
Here the PR to revert: https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/924

**Comment by mysticaltech** [comment]
@M4t7e Perfect, thanks for that, merged and preparing a quick release now!

**Reply by mysticaltech** [reply, marked as answer]
@grig0701 @MoergJ  @Keyruu FYI, @M4t7e's fix has just been released in v2.5.2. Please test and let us know. Follow the readme well and if you have another issue, please open an issue and share your kube.tf.

**Reply by MoergJ** [reply]
I was just setting up and fixing, @M4t7e beat me to it I guess, thank you 😆

**Reply by M4t7e** [reply]
@MoergJ Sorry for that :pray:  I was on it after my first comment when I tried to reproduce. Didn't see you were asked for a PR.

**Reply by MoergJ** [reply]
all good, thanks for the fix 😄

**Comment by mysticaltech** [comment]
@mnbro FYI, we had to revert #908 completely as it was causing problems. Whatever problem you had with it hanging in your case, please do open an issue with your kube.tf for us to reproduce and debug it further.

**Reply by mysticaltech** [reply]
@aleksasiriski just FYI. We both wrongfully approved that PR  😅

**Comment by grig0701** [comment]
@mysticaltech @M4t7e  

Yes, it works! Thank you very much for your quick response and correction.

Thank you for this beautiful module)

**Comment by grig0701** [comment]
@MoergJ thanks for the help)
